tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright (c) 2019, Lawrence Livermore National Security, LLC
+and DESUL project contributors. See the COPYRIGHT file for details.
+Source: https://github.com/desul/desul
+
+SPDX-License-Identifier: (BSD-3-Clause)
+*/
+
+#ifndef DESUL_ATOMICS_COMPARE_EXCHANGE_OPENACC_HPP_
+#define DESUL_ATOMICS_COMPARE_EXCHANGE_OPENACC_HPP_
+
+#include <openacc.h>
+
+#include <desul/atomics/Common.hpp>
+#include <desul/atomics/Thread_Fence_OpenACC.hpp>
+
+namespace desul {
+namespace Impl {
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<!std::is_arithmetic<T>::value, T> device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  //FIXME_OPENACC OpenACC lock APIs are not implemented.
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#ifdef KOKKOS_COMPILER_NVHPC
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_arithmetic<T>::value && (sizeof(T) != 4) && (sizeof(T) != 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_integral<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(unsigned int) == 4,
+                "this function assumes an unsigned int is 32-bit");
+  unsigned int return_val = atomicExch(reinterpret_cast<unsigned int*>(dest),
+                                       reinterpret_cast<unsigned int&>(value));
```

Comment:
So the compiler will emit host code when appropriate?

Reply:
CUDA allows the same atomic intrinsic functions to be called on both the host and the device; therefore, the same CUDA atomic-intrinsic-based implementation is used for both the host and the device.